swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/LogCompressor.swift

```diff
@@ -9,91 +9,116 @@ import AppleArchive
-      let logFilesFolderPath = FilePath(logFilesFolderURL)
+  public func start(
+    source directory: URL,
+    to file: URL
+  ) throws {
+    guard let path = FilePath(file),
+          let stream = ArchiveByteStream.fileStream(
+            path: path,
+            mode: .writeOnly,
+            options: [.create],
+            permissions: FilePermissions(rawValue: 0o644)
+          )
     else {
-      throw SettingsViewError.logFolderIsUnavailable
+      throw CompressionError.unableToOpenWriteStream
     }
 
-    let fileManager = FileManager.default
-    let fileURL =
-      destinationURL
-      ?? fileManager.temporaryDirectory.appendingPathComponent(fileName)
-
-    // Remove logfile if it happens to exist
-    try? fileManager.removeItem(at: fileURL)
-
-    // Create the file stream to write the compressed file
-    guard let filePath = FilePath(fileURL),
-      let writeFileStream = ArchiveByteStream.fileStream(
-        path: filePath,
-        mode: .writeOnly,
-        options: [.create],
-        permissions: FilePermissions(rawValue: 0o644))
+    try compress(source: directory, writeStream: stream)
+  }
+
+  public func start(
+    source directory: URL,
+    to byteStream: TunnelArchiveByteStream
+  ) throws {
+    guard let stream = ArchiveByteStream.customStream(
+      instance: byteStream
+    )
     else {
-      Log.app.error("\(#function): Couldn't create the file stream")
-      return nil
-    }
-    defer {
-      try? writeFileStream.close()
+      throw CompressionError.unableToOpenWriteStream
     }
 
-    // Create the compression stream
-    guard
-      let compressStream = ArchiveByteStream.compressionStream(
-        using: .lzfse,
-        writingTo: writeFileStream)
+    try compress(source: directory, writeStream: stream)
+  }
+
+  // Compress to a given writeStream which was opened either from a FilePath or
+  // TunnelArchiveByteStream
+  private func compress(
+    source directory: URL,
+    writeStream: ArchiveByteStream
+  ) throws {
+    let compressionStream = try openCompressionStream(writeStream)
+    let encodeStream = try openEncodeStream(compressionStream)
+    let keySet = try defineHeaderKeys()
+
+    guard let sourcePath = FilePath(directory)
     else {
-      Log.app.error("\(#function): Couldn't create the compression stream")
-      return nil
-    }
-    defer {
-      try? compressStream.close()
+      throw CompressionError.unableToReadSourceDirectory
     }
 
-    // Create the encoding stream
-    guard let encodeStream = ArchiveStream.encodeStream(writingTo: compressStream) else {
-      Log.app.error("\(#function): Couldn't create encoding stream")
-      return nil
-    }
-    defer {
-      try? encodeStream.close()
+    try encodeStream.writeDirectoryContents(
+      archiveFrom: sourcePath,
+      keySet: keySet
+    )
+
+    try? encodeStream.close()
+    try? compressionStream.close()
+    try? writeStream.close()
```

Comment:
Does Swift have `defer` statements or some other kind of RAII behaviour? Will these be closed in case the function throws an exception?